enhancements/etcd/automated-defrag.md

```diff
@@ -0,0 +1,235 @@
+free page and an accounting of all free pages are stored in a freelist for later reuse.
+
+Compaction does reduce the memory footprint by removing old keys but free pages can not be reclaimed back to disk
+without defragmentation.
+
+## Defrag Controller
+
+During the operator sync the controller will make a  call to the cluster API using the `MemberListRequest` RPC which 
+returns the current list of etcd members. We will then loop these members and for each dial the maintenance API using
+the `StatusRequest` RPC which will return detailed information on the backend status.
+
+```go
+type StatusResponse struct {
+    // dbSize is the size of the backend database physically allocated, in bytes, of the responding member.
+    DbSize int64 `protobuf:"varint,3,opt,name=dbSize,proto3" json:"dbSize,omitempty"`
+    [...] 
+    // dbSizeInUse is the size of the backend database logically in use, in bytes, of the responding member.
+    DbSizeInUse int64 `protobuf:"varint,9,opt,name=dbSizeInUse,proto3" json:"dbSizeInUse,omitempty"`
+}
+```
+
+From these two fields we can conclude the total fragmented space on that etcd members backend store. If the
+percentage of bytes of the total keyspace are over a defined threshold percentage. The controller will 
+utilize maintenance API again and issue a `DefragmentRequest` RPC.
+
+Defragmentation is I/O blocking thus important to only defrag a single etcd at a time. Additionally, to reduce
+the possibility of multiple leader elections we defragment the leader last. The amount of time that is required to
+defragment the store is directly related to the amount of free pages that exists. For this reason we feel it
+makes sense to continuously observe the clusters state and defrag frequently along the cadence of 
+compaction resulting in shorter intervals of disruption by keeping the number of pages limited and the size of the 
+database smaller.
+
+The criteria for defragmentation is defined as:
+
+- Cluster must be observed as `HighlyAvailableTopologyMode`.
+- The etcd cluster must report all members as healthy.
+- `minDefragBytes` this is a minimum database size in bytes before defragmentation should be considered.
+- `maxFragmentedPercentage` this is a percentage of the store that is fragmented and would be reclaimed after
+  defragmentation.
+  
+TODO show performance graphs.
+
+### Risks and Mitigations
+
+- SNO: As defragmentation is I/O blocking it is unsafe for non HA clusters. Mitigation to this risk includes a check in 
+  the controller for control-plane topology and will exit 0 if non HA.
+  
+- datafile corruption: Recently a bug was fixed in upstream etcd[1] which could lead to corruption of the key space 
+  as a result of defragmentation if it happened while etcd was being terminated. Mitigation to this risk includes 
+  heavy exercise in CI and telemetry will alert `etcdMemberDown` if one of the etcd instances goes down as the 
```

Comment:
etcdMemberDown fires after 1 minute, is this short or long enough for this?

Reply:
the timeout for the call is 45s technically we could see a wait that long but we would miss 1m.